### PR TITLE
Update list.{count, index, remove} to accept `object` type

### LINKED ISSUE #15271: Several methods on built-in types are too strict.
I have identified several methods on built-in types that are too strict and can cause annoying false positives:

- [x] set and frozenset methods   (#15470, #15160)
- [ ] `dict.{get, pop}` (#15471)
- [ ] `list.{count, index, remove}` (#15472)
- [ ] `MutableSet.discard`
- [ ] `Mapping.get`
- [ ] `MutableMapping.pop` 
- [ ] `Sequence.{count, index}`
- [ ] `MutableSequence.remove`
- [ ] probably some more.

It seems like the annotations for these were designed to help detect user errors. For instance, we know that `.get` on a `dict[int, int]` will always return the default as the types are incompatible, and so `.get` was restricted in typeshed to only accept the key type. However, this hurts, even in simple cases like using literals:

```python
from typing import Mapping, Literal

def demo_get(
    d: Mapping[Literal["foo", "bar"], int],
    dynamic_key: str,
) -> None:
    # No overload variant of "get" of "Mapping" matches argument type "str"
    d.get(dynamic_key)  

def demo_sub(
    left: set[Literal["foo", "bar"]],
    right: set[str],
) -> None:
    # Unsupported operand types for - ("set[Literal['foo', 'bar']]" and "set[str]")
    left - right
```

I believe typeshed is trying to act as a linter here, which shouldn't be its job. Type-checkers and type-informed linters are better at this and can use more complex logic.

For example, in the set difference case `set[A] - set[B]`, the proper test would be to check whether the types have a non-empty intersection or not. The same could be said for `Mapping[K, V].get(K2)`.

## Related Topics:

- https://github.com/python/typeshed/issues/8219
- https://github.com/python/typeshed/issues/9155
- https://github.com/python/typeshed/issues/9004

### PR DESCRIPTION
**Pull request by randolf-scholz**
Partially fixes #15271

### CODE REVIEW

**Comment by randolf-scholz** [comment]
Hm, not exactly sure what's going on with `numpy`, my best guess is that since `_NestedSequence` is defined with (https://github.com/numpy/numpy/blob/main/numpy/_typing/_nested_sequence.py)

```python
    def count(self, value: Any, /) -> int:
        """Return the number of occurrences of `value`."""
        raise NotImplementedError

    def index(self, value: Any, /) -> int:
        """Return the first index of `value`."""
        raise NotImplementedError
```

then after we make `list` use `object` rather than `Any`, this is still a subtype but possibly no longer a "`proper_subtype`". This is a mypy interal distinction and likely needs to be addressed on mypys side. 

EDIT: I am starting to think that this may be related to mypy interal solver semantics. we removed some constrains, so now it may find solutions for the `T` of a list-constructor inside nested function calls, where previously there was no solution and it defaulted to `Never` which then propagate throught the call stack and may change inference in some cases.

**Comment by srittau** [comment]
The primer output is not promising. The fixed `# type: ignore`s are mostly in test code. On the other hand there seems to be some churn, and even some true positives (in `colour`) that aren't flagged anymore.

**Comment by randolf-scholz** [comment]
The main issue seems to be [var-annotated] issue originating from numpy:


```python
import numpy as np

reveal_type(np.array([]))
```

```
# on main
tmp_a.py:3: note: Revealed type is "numpy.ndarray[tuple[Any, ...], numpy.dtype[Any]]"
# PR:
tmp_a.py:3: note: Revealed type is "numpy.ndarray[tuple[Any, ...], numpy.dtype[Never]]"
```

**Comment by randolf-scholz** [comment]
@srittau  I wouldn't call the colour results true-positives. For instance https://github.com/colour-science/colour/blob/c9cfa1f333452cc05e42f09eebd2f3dd935a5981/colour/utilities/network.py#L1756-L1773 is exactly one of the cases of a valid `list.remove` call that is currently a false positive.

**Comment by randolf-scholz** [comment]
OK, I think I have a repro of the numpy issue that causes the [var-annotated] errors. It has to do with the [`NestedSequence`](https://github.com/numpy/numpy/blob/4d711763f8c5b978abd88cea710147c14d802675/numpy/_typing/_nested_sequence.py#L12) Protocol. Here is a minimal reproduction:


```python
from typing import Any, Protocol, reveal_type

class NestedSeq[_T_co](Protocol):

    def __getitem__(self, index: int, /) -> "_T_co | NestedSeq[_T_co]":
        raise NotImplementedError

    def count(self, value: Any, /) -> int:
        raise NotImplementedError

def as_nested_sequence[T](seq: NestedSeq[T], /) -> NestedSeq[T]:
    return seq

reveal_type(as_nested_sequence([]))
# main: NestedSeq[Any], PR: NestedSeq[Never]
```

And this happens due to how mypy solves type-vars. on main, `count` is `[T] -> int`, so mypy solves `T=Any`, whereas on PR, `count` is `[object] -> int`, so mypy has no constraints for `T` and picks `T=Never`.

Imo, this is more of a mypy problem than a typeshed problem. But it could be worked around if for example numpy used `default=Any` for the type-variable of `_NestedSequence`. cc @jorenham

**Comment by jorenham** [comment]
> Imo, this is more of a mypy problem than a typeshed problem. But it could be worked around if for example numpy used `default=Any` for the type-variable of `_NestedSequence`. cc @jorenham

Sure, we could do that.

**Comment by jorenham** [comment]
> > Imo, this is more of a mypy problem than a typeshed problem. But it could be worked around if for example numpy used `default=Any` for the type-variable of `_NestedSequence`. cc @jorenham
> 
> Sure, we could do that.

The workaround will be present in the upcoming 2.4.5 and 2.5.0 releases.

**Comment by jorenham** [comment]
numpy 2.4.5 was just released (https://github.com/numpy/numpy/releases/tag/v2.4.5)

**Comment by randolf-scholz** [comment]
@jorenham  Ugh, terribly sorry but that wasn't sufficient. `mypy` still resolves `_ScalarT` to `Never` in this overload:

```python
@overload
def array(
    object: _ArrayLike[_ScalarT],
    dtype: None = None,
    *,
    copy: bool | _CopyMode | None = True,
    order: _OrderKACF = "K",
    subok: bool = False,
    ndmin: int = 0,
    ndmax: int = 0,
    like: _SupportsArrayFunc | None = None,
) -> NDArray[_ScalarT]: ...
```

because it doesn't find any constraints on `_ScalarT` when doing `np.array([])`, so mypy emits `var-annotated`[^1].

So actually it is this typevar that needs a default (either `Any` or `np.generic` I guess).

Again, sorry for not testing this more thoroughly locally, doing the default on `_NestedSequence` itself in hindisght obviously only helps when we use `_NestedSequence` directly, but not with `def [T](_NestedSequence[T]) -> Array[T]` type constructors. Then we need `def [T=Any](_NestedSequence[T]) -> Array[T]`  instead.

[^1]: Which, let's be honest, probably shouldn't even be an error enabled by default. `pyrights` / `pyrefly` / `ty` all use `list[Unknown]` when doing `x = []` and don't emit any errors with standard settings.

**Comment by jorenham** [comment]
Ah no worries. At least we know what to put in the mypy bug report now ;)

**Comment by randolf-scholz** [comment]
@srittau I inspected the `mypy-primer` in more detail. Looking at the issues in detail, I believe the `mypy-primer` changes are positive in almost all cases.

## optuna

[visualization_tests/test_rank.py#L699](https://github.com/optuna/optuna/blob/a382306b13afba28515ff132e13cab2a743c5729/tests/visualization_tests/test_rank.py#L699): This error should be expected. If `mypy` complains on `x2=[]`, then one should expect it to also complain on `x2=np.array([])`

## colour

- [colour/utilities/network.py:1772](https://github.com/colour-science/colour/blob/c9cfa1f333452cc05e42f09eebd2f3dd935a5981/colour/utilities/network.py#L1772) This is a false positive. Here, `node` is an instance of `PortNode` and `_children` is annotated as `list[Self]`, i.e. `list[PortGraph]`, and `PortGraph` subclasses `PortNode`. Since `list.remove` is equality based, this does the right thing at runtime, and shows that the current annotation is too strict.

- [`colour/notation/munsell/centore2014.py:2053`](https://github.com/colour-science/colour/blob/c9cfa1f333452cc05e42f09eebd2f3dd935a5981/colour/notation/munsell/centore2014.py#L2053) the flagged code here is correct. The underlying issue is an incorrect annoation in `_munsell_maximum_chromas_from_renotation`,  is annotated as returning `tuple[tuple[Pair, float]]` , but actually returns `tuple[tuple[Triple, float]]` (with `Triple=tuple[float, float, float]`, `Pair=tuple[float, float]`). `mypy` doesn't catch this incorrect annotation because `chromas = {}` is not annotated inside the function.


## pandas

- the `var-annotated` errors are expected, as with the `optuna` case.
- [pandas/plotting/_matplotlib/core.py:1207](https://github.com/pandas-dev/pandas/blob/9993ce6161342306500378d44700daccbd9f7aa3/pandas/plotting/_matplotlib/core.py#L1207) Not enitrely sure what happens here, didn't debug this one.

## xarray

- [xarray/tests/test_concat.py:112](https://github.com/pydata/xarray/blob/8bb033ceb607558e13b108885ba866b6621e3d2e/xarray/tests/test_concat.py#L112) Not exactly sure what's happening here, if required I could look into it
- [xarray/tests/test_backends.py:2090](https://github.com/pydata/xarray/blob/8bb033ceb607558e13b108885ba866b6621e3d2e/xarray/tests/test_backends.py#L2090) I am not entirely sure what happens here, but it seems that even on `main`, `np.ma.array(y, mask=[True, True, False, False, False])` with `y: list[float]` produces a `var-annotated`. However, with an inlined list constructor, on main `mypy` thinks that the first list expression in `np.ma.array([-1, -1, 10, 10.1, 10.2], mask=[True, True, False, False, False])` is a `list[Any]`, whereas with this PR it thinks it is a `list[float]`, which is more accurate.

## scipy-stubs

- [tests/linalg/test_blas.pyi:27](https://github.com/scipy/scipy-stubs/blob/5e766f020adfa98da34015a550416d78dbcf0a67/tests/linalg/test_blas.pyi#L27) With the new annotations, `list[str]` should satisfy the `_SequenceNotStr[str]` Protocol, which it previously didn't. This overload and result seems more precise then bevor.
- 3 unused-ignores, nice
- [tests/special/test_ufuncs.pyi:75](https://github.com/scipy/scipy-stubs/blob/5e766f020adfa98da34015a550416d78dbcf0a67/tests/special/test_ufuncs.pyi#L75) This is already flagged with `pyright: ignore` and `pyrefly: ignore`, so I'd assume its a mypy false-negative.

**Comment by randolf-scholz** [comment]
The pyright failures seem unrelated

**Comment by jorenham** [comment]
> * [tests/special/test_ufuncs.pyi:75](https://github.com/scipy/scipy-stubs/blob/5e766f020adfa98da34015a550416d78dbcf0a67/tests/special/test_ufuncs.pyi#L75) This is already flagged with `pyright: ignore` and `pyrefly: ignore`, so I'd assume its a mypy false-negative.

Those are indeed tests for true positives. The trick is that in scipy-stubs all type-checkers are configured to complain in case of unused ignore comments. So it's indeed a good thing that mypy now also reports an error.

**Comment by Akuli** [comment]
I think `list[int].index(12.34)` should be a type checker error. If the thing can fail at runtime with some float values, we shouldn't say that `float` is supported.

More generally, I don't think it makes sense to allow an overlapping type in these methods (let alone `object`). If the runtime returns a fallback or is otherwise silent, like `dict.get` or `set.discard`, that's different.

**Comment by randolf-scholz** [comment]
@Akuli There are legitimate use cases, for instance calling `.index` with a union type (like `T | None`) or supertype of the generic type.

```python
class RGB: ...
class RGBA(RGB): ...

class Demo:
    colors: list[RGBA]

    def check_value(self, value: RGB) -> None:
        if value in self.colors:
            position = self.colors.index(value)
            ...
```

I understand why you want `list[int].index(12.34)` to be a typing error, this is a code smell and likely incorrect code. But again, if we set this in stone in the stubs you deny any legitimate use case. What you want should imho be some opt-in linter rule that type checkers like `mypy` or `ty` can implement if they want. But `typeshed` is the wrong place for this imo.

**Comment by Akuli** [comment]
In your example, ideally the type checker would narrow the type of `value` into `RGBA` because it is in a `List[RGBA]`. Then the `.index()` would succeed as is.

You could also make the same argument with any other operation that needs an `RGBA` object, and I don't see how `list.index()` is special here. For example:

```python
class RGB:
    red: float
    green: float
    blue: float
    ...

class RGBA(RGB):
    alpha: float
    ...

colors: list[RGBA] = [...]
value = get_some_rgb_color()
if value in colors:
    print(f"opacity is {value.alpha}")
```

More generally, the correct fix is to help type checkers know the correct type, not to silence errors when they have the wrong type.

To be concrete, I would be very surprised if mistakes like `list[int].index(12.34)` (or `list[int].index("hello")`) were less frequently occurring, and hence less important, than the problems that this PR solves. And as I said in your other PR, mypy_primer is not a good way to measure this because most of the code it checks already passes the type check.

**Comment by randolf-scholz** [comment]
> In your example, ideally the type checker would narrow the type of value into RGBA because it is in a List[RGBA]. Then the .index() would succeed as is.

This would not be ideal, but actually incorrect as list containment is equality based. Your own example will error on the line `print(f"opacity is {value.alpha}")` if `RGBA` overwrites `__eq__` in such a way that an RGB can equal an RGBA:

```python
from dataclasses import dataclass

@dataclass(frozen=True)
class RGB:
    red: int
    green: int
    blue: int

@dataclass(frozen=True)
class RGBA(RGB):
    alpha: int = 255

    def __eq__(self, other):
        if isinstance(other, RGBA):
            return hash(self) == hash(other)
        if isinstance(other, RGB):
            return RGB.__hash__(self) == hash(other) and self.alpha == 255
        return NotImplemented

black = RGB(0, 0, 0)
colors = [RGBA(0, 0, 0)]
if black in colors:
    print(black.alpha)  # ERROR at runtime
```

another example with pytorch:

```python
import torch

# Note: torch.Size subclasses tuple[int, ...]
sizes: list[torch.Size] = [torch.Size([1, 2, 3])]
print(sizes[0].numel())  # 6 (product of values)
value = (1, 2, 3)
if value in sizes:
    print(value.numel())  # ERROR
```

**Comment by randolf-scholz** [comment]
There is actually a regression in `mypy` type checking here, introduced in 1.20.0 https://github.com/python/mypy/issues/21512

**Comment by Akuli** [comment]
I see. There might not be a way to express this so that overrided `__eq__` methods and the error for `list[int].index(float)` both work. But even then, falling back to `object` is just not the way to go.

There are probably many other places too where overrided `__eq__` methods behave weirdly.

**Comment by randolf-scholz** [comment]
Again I think what you want is a fine strict linter rule to have, but you seem very eager to force this linter rule on everyone else, even when it caues false positives in their code.

**Comment by jorenham** [comment]
don't worry about the scipy-stubs mypy_primer diff
